lib/web_ui/lib/src/engine/keyboard_binding.dart

```diff
@@ -0,0 +1,489 @@
+  // [dispatchKeyData].
+  //
+  //  * The method might dispatch some synthesized key data first to update states,
+  //    results discarded.
+  //  * Then it dispatches exactly one non-synthesized key data that corresponds
+  //    to the `event`, i.e. the main key data. The result of this dispatching is
+  //    returned to indicate whether the event is processed by Flutter.
+  //  * Some key data might be synthesized to update states after the main key
+  //    data. They are always scheduled asynchronously with results discarded.
+  bool handleEvent(FlutterHtmlKeyboardEvent event) {
+    final Duration timeStamp = _eventTimeStampToDuration(event.timeStamp!);
+
+    if (_shouldPreventDefault(event)) {
+      event.preventDefault();
+    }
+    final String eventKey = event.key!;
+
+    final int physicalKey = _getPhysicalCode(event.code!);
+    final bool logicalKeyIsCharacter = !_eventKeyIsKeyname(eventKey);
+    final String? character = logicalKeyIsCharacter ? eventKey : null;
+    final int logicalKey = () {
+      if (kWebLogicalLocationMap.containsKey(event.key!)) {
+        final int? result = kWebLogicalLocationMap[event.key!]?[event.location!];
+        assert(result != null, 'Invalid modifier location: ${event.key}, ${event.location}');
+        return result!;
+      }
+      if (character != null)
+        return _characterToLogicalKey(character);
+      if (eventKey == _kLogicalDead)
+        return _deadKeyToLogicalKey(physicalKey, event);
+      return _otherLogicalKey(eventKey);
+    }();
+
+    assert(event.type == 'keydown' || event.type == 'keyup');
+    final bool isPhysicalDown = event.type == 'keydown' ||
+      // On macOS, both keydown and keyup events of CapsLock should be considered keydown,
+      // followed by an immediate cancel event.
+      (_shouldSynthesizeCapsLockUp() && event.code! == _kPhysicalCapsLock);
+
+    final int? lastLogicalRecord = _pressingRecords[physicalKey];
+
+    ui.KeyChange change;
+
+    if (_shouldSynthesizeCapsLockUp() && event.code! == _kPhysicalCapsLock) {
+      // Case 1: Handle CapsLock on macOS
+      //
+      // On macOS, both keydown and keyup events of CapsLock are considered
+      // keydown, followed by an immediate synchronized up event.
+
+      _scheduleAsyncEvent(
+        Duration.zero,
+        () => ui.KeyData(
+          timeStamp: timeStamp,
+          change: ui.KeyChange.up,
+          physical: physicalKey,
+          logical: logicalKey,
+          character: null,
+          synthesized: true,
+        ),
+        () {
+          _pressingRecords.remove(physicalKey);
+        }
+      );
+      change = ui.KeyChange.down;
+
+    } else if (isPhysicalDown) {
+      // Case 2: Handle key down of normal keys
+      change = ui.KeyChange.down;
+      if (lastLogicalRecord != null) {
+        // This physical key is being pressed according to the record.
+        if (event.repeat ?? false) {
```

Comment:
Another pattern that I've seen is `if (event.repeat == true) { ... }`. I couldn't find any recommendation in the style guide, so I'm not sure which patter is preferred.

Reply:
Thanks for pointing out, although I do think this style is more readable.